Sources/ApodiniDeployBuildSupport/Options/Namespace.swift

```diff
@@ -0,0 +1,28 @@
+//
+//  Namespace.swift
+//
+//
+//  Created by Lukas Kollmer on 2021-02-06.
+//
+
+import Foundation
+
+
+
+public protocol OuterNamespace: class {
+    static var id: String { get }
+}
+
+//extension OuterNamespace {
+//    static var id: String { "\(Self.self)" }
+//}
+
+
+public protocol InnerNamespace: class {
+    associatedtype OuterNS: OuterNamespace
+    static var id: String { get }
+}
```

Comment:
If I understand this correctly the inner namespace is used e.g. to specify deployment options that are specific to one deployment provider and the generic options also have their own inner namespace that is used for them? Maybe renaming it to `Namespace` and then `Category` or `Subset` or something like this might be more fitting?

Reply:
Yes pretty much. I tried to design the Options API as abstract as possible, hence the two-level namespacing.
The outer namespace creates scope for options, and the inner namespace is used to differentiate between otherwise-duplicate options within that scope.
Currently we have only one outer namespace (the one for deployment options), but in theory if eg we wanted to move the `@Parameter` options to this API, we could simply define a second outer namespace for ParameterOptions, and there would be a type-system-level check to make sure you cannot pass a deployment option to a parameter and vice versa.

I'll see what i can do about the naming. I agree that the Outer/InnerNamespace names aren't exactly perfect, but tbh I'm not sure `Category` or `Subset` are much better, since the inner namespace neither defines a category, nor a subset.

Reply:
I think moving the `Parameters` over to the new Options API would be a good idea, I like that you already had this extensibility in mind 👍
I agree that `Outer`/`InnerNamespace` or `Category` and `Subset` are not the best names, maybe we can find a better name along the way.

Reply:
Yes I think that might work, although you'd lose some convenience because you's have to manually specify keys instead of using pointer equality to compare options (which is how the parameter options API currently works).

I'll have a look.